### ISSUE TITLE
Why?

### ISSUE THREAD

**Issue by RX14** [firejox/crystal-async_await#1]
If this is just a learning exercise, that's great, but I can't see the advantage in using this for real projects. Take for example the example in the readme, it can be written using normal crystal simply (except for sleep) by *removing* syntax. Is there any advantages to this approach I'm missing versus Crystal's fibers and channels?

```cr
require "async_await"

class Foo
  getter ch = AAChannel(String).new

  async def bar
    await Task.delay(Time::Span.new(0, 0, 1))
    await ch.receive  # => "Hello"
    await ch.send "Crystal!"
  end
end

foo = Foo.new

async_spawn do
  foo.bar
end

foo.ch.send_with_csp "Hello"
foo.ch.receive_with_csp # => "Crystal!"
```

```cr
class Foo
  getter ch = Channel(String).new

  def bar
    sleep 1
    ch.receive  # => "Hello"
    ch.send "Crystal!"
  end
end

foo = Foo.new

spawn do
  foo.bar
end

foo.ch.send "Hello"
foo.ch.receive # => "Crystal!"
```

**Comment by firejox** [comment]
Currently, Crystal's fiber and channel can't work fine in multi-thread.  This repo is focus on handling multi-thread. Yeah, I agree that every thing that  **async/await** can do, fiber can also do.
And **async/await** is just one of features, you can use it without any **async/await**. But you
can't use Crystal's channel without fiber.

**Comment by firejox** [comment]
@RX14 Actually, It can combine future and channel together.
```Crystal
require "async_await"

class Foo
  getter ch = AAChannel(Int32).new

  async def bar
    task = @ch.receive
    bar2
    p x = await task
    x + 1
  end

  def bar2
    p 1
  end
end

foo = Foo.new

x = foo.bar
foo.ch.send_with_csp 2
p x.value_with_csp
```
the above code would output
```
1
2
3
```
you can evaluate the value when you want, not just only channel receive or method return

**Comment by RX14** [comment]
Futures exist in crystal, built on channels.

```cr
class Foo
  getter ch = Channel(Int32).new

  def bar
    task = @ch.receive
    bar2
    p x = task
    x + 1
  end

  def bar2
    p 1
  end
end

foo = Foo.new

x = future { foo.bar }
foo.ch.send 2
p x.get
```

I believe that getting blocking behaviour by default, and explicitly asking for concurrency is better than littering every function with `async`/`await`.

**Comment by firejox** [comment]
Uh...your code a little bit different to me
If you insert `p "HI"` after foo.bar, you will get different output

```Crystal
require "async_await"

class Foo
  getter ch = AAChannel(Int32).new

  async def bar
    task = @ch.receive
    bar2
    p x = await task
    x + 1
  end

  def bar2
    p 1
  end
end

foo = Foo.new

x = foo.bar
p "HI"
foo.ch.send_with_csp 2
p x.value_with_csp
```
the output is
```
1
"HI"
2
3
```
but yours
```Crystal
class Foo
  getter ch = Channel(Int32).new

  def bar
    task = @ch.receive
    bar2
    p x = task
    x + 1
  end

  def bar2
    p 1
  end
end

foo = Foo.new

x = future { foo.bar }
p "HI"
foo.ch.send 2
p x.get
```
it will get
```
"HI"
1
2
3
```

**Comment by RX14** [comment]
@firejox Does it really matter?

**Comment by firejox** [comment]
@RX14 Yes. If you want to do the same thing, you would write some scary future like that.
```Crystal
class Foo
  getter ch = Channel(Int32).new

  def bar
    task = future { @ch.receive }
    bar2
    future do
      p x = task.get
      x + 1
    end
  end

  def bar2
    p 1
  end
end

foo = Foo.new

x = foo.bar
p "HI"
foo.ch.send 2
p x.get
```
The `async/await` can prevent this situation.
Actually, this repo is mainly on multi-thread handling. It can be a bridge between threads and fibers since current Cystal's fiber and channel can't do well in multi-thread.

**Comment by davy-tw** [comment]
`async`/`await` pattern is also exists in ES7 (JavaScript), [Python 3](https://docs.python.org/3/library/asyncio-task.html), and [C#](https://msdn.microsoft.com/library/hh191443(v=vs.110).aspx).

It can build async methods easily with simple syntax, when executes async methods, the language will automatically create a new thread (maybe fiber or uses thread pool here, it doesn't matter) for this async method call (for parallel-executing).

When we are using a lib with async version methods, we can simply call these method without creating/maintaining threads/fibers by ourself, and if we need to run it in-order, we can just use `await` syntax to wait all progresses in right-hand expression end.

Yes, we can use the Crystal-builtin ways to do the same things, but this is still a good syntax here.

**Comment by RX14** [comment]
We could probably trade examples all day. The two syntaxes end up very similar in most cases. Personally, I prefer CSP to async/await because I find using async/await keywords everywhere tiring. I also think it's a better abstraction because you build promises on top of sequential execution instead of building sequential execution on top of promises which are built on callbacks. Adding `future { }` around things occasionally seems to me better than having async/await keywords everywhere.

But even if you regard both methods as being exactly equal, you're still fighting against the standard library and the whole ecosystem, which are already built on CSP. If you want to be multi-threaded why not just [contribute to crystal](https://github.com/crystal-lang/crystal/tree/thread-support)?

**Comment by firejox** [comment]
Well, I think that would change a lot in standard library even if I contribute the channel part 
to Crystal. It would really challenge the Crystal's community. And this repo is still incomplete.
I can't contribute that into standard library.

In fact, I cannot figure out what the theoretical difference between `CSP` and `async/await`.
I think `async/await` is just the one of part in `CSP`.
